### PR TITLE
Implemented extracting article text and displaying it

### LINKED ISSUE #4: Extract Text from Article
use the browser extension capabilities to capture the article text of current webpage

### Acceptance Criteria:
- be able to extract main article text
- should be tested to remove unrelated HTML tags and meta info

### PR DESCRIPTION
**Pull request by sauravbanna**
Resolves #4 

Currently using HTML parsing + regex to get text and get rid of <a> tags, etc.

### CODE REVIEW

File: backend/get_article_text.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+            const all_text = [];
+
+            for (const text_element of text_elements) {
+                const span_regex = /<span.*>/g
+                const a_tag_regex = /<\/?a([^>])*>/g
+
+                const text = text_element.innerHTML;
+
+                if (text.charAt(0) != '<' && !(span_regex.exec(text))) {
```
Comment (Michaeltaoma) [inline backend/get_article_text.js:32]: consider using strict equality operator

**Comment by Michaeltaoma** [comment]
Good job! When I tested the extension, I found that the extension would not go back to display the get article option when I switched to another news article. The only way to return to the main menu is to click the extension again. We can make it automatically go back to the "main menu" when switching articles.

**Comment by Michaeltaoma** [comment]
Also we might want to unify our naming conventions.

File: backend/package.json [hunk 2]
```diff
@@ -1,8 +1,11 @@
 {
   "dependencies": {
+    "@mozilla/readability": "^0.4.2",
     "@nlpjs/lang-en": "^4.23.4",
     "chai": "^4.3.7",
     "cors": "^2.8.5",
+    "express": "^4.18.2",
+    "jsdom": "^20.0.2"
```
[the remark this answers is not in the document]
Comment (sauravbanna) [reply backend/package.json]: done

**Comment by armintalaie** [comment]
@sauravbanna @Michaeltaoma correct me if I'm wrong, but the text extraction is a light task that the frontend (in this case the extension) can do on its own. the code for extraction that's in the backend can be moved to the frontend. we can keep the endpoint since for summarizing we will end up passing the text anyway but "parsing" it would be a waste of server resources. this isn't an issue for testing but a scenario where numerous calls are made to the server can use up a lot of its power and affect performance too.

**Comment by sauravbanna** [comment]
> @sauravbanna @Michaeltaoma correct me if I'm wrong, but the text extraction is a light task that the frontend (in this case the extension) can do on its own. the code for extraction that's in the backend can be moved to the frontend. we can keep the endpoint since for summarizing we will end up passing the text anyway but "parsing" it would be a waste of server resources. this isn't an issue for testing but a scenario where numerous calls are made to the server can use up a lot of its power and affect performance too.

That makes sense, so the same code just without the server call right? I'll change that today

**Comment by Michaeltaoma** [comment]
>

**Comment by Michaeltaoma** [comment]
> @sauravbanna @Michaeltaoma correct me if I'm wrong, but the text extraction is a light task that the frontend (in this case the extension) can do on its own. the code for extraction that's in the backend can be moved to the frontend. we can keep the endpoint since for summarizing we will end up passing the text anyway but "parsing" it would be a waste of server resources. this isn't an issue for testing but a scenario where numerous calls are made to the server can use up a lot of its power and affect performance too.

I agree that the "extracting text from the article" is trivial enough that FE can handle the job. However, we could have a clearer boundary for "trivial" and "non-trivial" to enforce a better distinction between the backend and frontend logic. From my perspective, BE should be doing all the logical processing (even some tasks that seem trivial right now) since it would be way better for us to build up the structure and not have things lying around so when the task becomes complicated in the future, it would be easier to adapt to it.

**Comment by armintalaie** [comment]
> > @sauravbanna @Michaeltaoma correct me if I'm wrong, but the text extraction is a light task that the frontend (in this case the extension) can do on its own. the code for extraction that's in the backend can be moved to the frontend. we can keep the endpoint since for summarizing we will end up passing the text anyway but "parsing" it would be a waste of server resources. this isn't an issue for testing but a scenario where numerous calls are made to the server can use up a lot of its power and affect performance too.
> 
> I agree that the "extracting text from the article" is trivial enough that FE can handle the job. However, we could have a clearer boundary for "trivial" and "non-trivial" to enforce a better distinction between the backend and frontend logic. From my perspective, BE should be doing all the logical processing (even some tasks that seem trivial right now) since it would be way better for us to build up the structure and not have things lying around so when the task becomes complicated in the future, it would be easier to adapt to it.

it's not a backend task tho. when you open an app and click on a button, you don't send a request to backend to ask it what button it pressed. the text extraction is literally UI work that we need to check the DOM and get the article. now what happens to the text can be sent to the backend. for example, summary, lens or any other thing.

**Comment by Michaeltaoma** [comment]
> > > @sauravbanna @Michaeltaoma correct me if I'm wrong, but the text extraction is a light task that the frontend (in this case the extension) can do on its own. the code for extraction that's in the backend can be moved to the frontend. we can keep the endpoint since for summarizing we will end up passing the text anyway but "parsing" it would be a waste of server resources. this isn't an issue for testing but a scenario where numerous calls are made to the server can use up a lot of its power and affect performance too.
> 
> > 
> 
> > I agree that the "extracting text from the article" is trivial enough that FE can handle the job. However, we could have a clearer boundary for "trivial" and "non-trivial" to enforce a better distinction between the backend and frontend logic. From my perspective, BE should be doing all the logical processing (even some tasks that seem trivial right now) since it would be way better for us to build up the structure and not have things lying around so when the task becomes complicated in the future, it would be easier to adapt to it.
> 
> 
> 
> it's not a backend task tho. when you open an app and click on a button, you don't send a request to backend to ask it what button it pressed. the text extraction is literally UI work that we need to check the DOM and get the article. now what happens to the text can be sent to the backend. for example, summary, lens or any other thing.

I see, got u!

**Comment by armintalaie** [comment]
@sauravbanna  please tweak and finalize your code + merge with main. The designs are ready and we can get started with those too

**Pull request merged**